### DISCUSSION TITLE
Using rust-attr with feature="serde" not working

### DISCUSSION THREAD

**Discussion by SKittan** [slint-ui/slint#3314] [Q&A] [answered]
I'm trying to test saving and loading a Slint struct by serde in Rust. I found a [pull request](https://github.com/slint-ui/slint/pull/2727#issuecomment-1549300108) about this topic and tested the code for my self.

The Slint file has only a small test struct and a window to display it:
```rust
@rust-attr(cfg_attr(feature="serde", derive(Serialize, Deserialize)))
export struct winData {
    number: int,
    text: string
}

export component mainWindow inherits Window {
    in-out property <winData> data;

    VerticalLayout {
        padding: 10px;
        Text {text: data.number;}
        Text {text: data.text;}
    }
}
```

The rust code is creating the window and loading a sample file:
```rust
use std::fs;

slint::slint!(import { mainWindow } from "./src/main.slint";);

fn main() {
    let main_window = mainWindow::new().unwrap();

    let fname = "../../data.json";
    match fs::read_to_string(&fname) {
        Err(_) => println!("Error open TMx devices file: {:?}",
                           &fname),
        Ok(data) => {
            match serde_json::from_str::<winData>(&data) {
                Ok(res) => {
                    main_window.set_data(res);
                },
                Err(e) => {
                    println!("Error reading TMx devices config: \
                             {:?}", e);
                }
        }}};

    main_window.run();
}
```

The compilation of this project fails with the following error:

> error[E0277]: the trait bound `winData: serde::de::Deserialize<'_>` is not satisfied
>     --> src\main.rs:13:19
>      |
> 13   |             match serde_json::from_str::<winData>(&data) {    
>      |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `serde::de::Deserialize<'_>` is not implemented for `winData`

**Comment by ogoffart** [comment]
Since your crate probably doesn't have a `serde` feature, you can try something like
`@rust-attr(derive(Serialize, Deserialize))`  (without the cfg_attr)

**Comment by SKittan** [comment]
This is the point I started from and leads to the following compiler error:

> error[E0277]: the trait bound `slint::SharedString: Deserialize<'_>` is not satisfied
    --> src\main.rs:3:1
     |
3    | slint::slint!(import { mainWindow } from "./src/main.slint"...
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Deserialize<'_>` is not implemented for `slint::SharedString`

**Comment by ogoffart** [comment]
ah, you need to enable the `serde` feature of the slint crate.

**Comment by SKittan** [comment]
Can you please provide a small example how to do this?

Edit:
To clarify my question: I can find the following feature flags in the Slint rust [docs](https://slint.dev/releases/1.1.1/docs/rust/slint/#feature-flags):
- compat-1-0
- std
- gettext
- libm
- log
- software-renderer-systemfonts
- unsafe-single-threaded
- accessibility

When I try `slint = { version = "1", features = ["serde"] }` in my cargo.toml file, I get an error. Do I understand something wrong or is the serde feature not available yet?

**Comment by SKittan** [comment, marked as answer]
In case someone else has the same challenge:

The `serde` feature of Slint can be activated with the following dependency entry in the `Cargo.toml` file:
`
i-slint-core = { version = "1.1", features = ["serde"] }
`

Afterwards the `serde` derive macros can be used in Slint files, as described in the documentation:
`
@rust-attr(derive(serde::Serialize, serde::Deserialize))
`

The activation of the serde feature will enable Serialization / Deserialization of the SharedString data type.

**Reply by ogoffart** [reply]
You don't want to add i-slint-core in your Cargo.toml because thats's private API and it can break

I thought we had a serde feature exposed in the slint crate, but apparently we did not add that. 
So i'm adding it in https://github.com/slint-ui/slint/pull/3326
